beacon-chain/rpc/over/handlers_deposit.go

```diff
@@ -0,0 +1,302 @@
+	if len(pds) == 0 {
+		return pdes, nil
+	}
+
+	finalizedEpoch := st.FinalizedCheckpointEpoch()
+	finalizedSlot, err := slots.EpochStart(finalizedEpoch)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get finalized slot")
+	}
+
+	depBalToConsume, err := st.DepositBalanceToConsume()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get deposit balance to consume")
+	}
+	availableForProcessing := depBalToConsume + helpers.ActivationBalanceChurnLimit(activeBalance)
+
+	currentEpoch := slots.ToEpoch(st.Slot())
+	processedAmount := uint64(0)
+	nextDepositIndex := uint64(0)
+
+	// state.pending_deposits is guaranteed to be sorted by slot, so we can iterate through it in order
+	// Mostly the same logic as in ProcessPendingDeposits, but it iterates through all pending deposits.
+	for _, pd := range pds {
+		// A pending deposit is processed when its slot is larger than finalized slot.
+		// In this case, moving to the next epoch is needed. This will update and initialize the variables.
+		// Premise: finalizedSlot will always be incremented by 32 slots(= 1 epoch).
+		for pd.Slot > finalizedSlot {
+			currentEpoch += 1
+
+			activeBalance += pendingBalanceToBeActivated[currentEpoch]
+			availableForProcessing = helpers.ActivationBalanceChurnLimit(activeBalance)
+			finalizedSlot += params.BeaconConfig().SlotsPerEpoch
+
+			processedAmount = uint64(0)
+			nextDepositIndex = uint64(0)
+		}
+
+		// Move to next epoch if max pending deposits per epoch is reached.
+		if nextDepositIndex >= params.BeaconConfig().MaxPendingDepositsPerEpoch {
+			currentEpoch += 1
+
+			activeBalance += pendingBalanceToBeActivated[currentEpoch]
+			availableForProcessing = helpers.ActivationBalanceChurnLimit(activeBalance)
+			finalizedSlot += params.BeaconConfig().SlotsPerEpoch
+
+			processedAmount = uint64(0)
+			nextDepositIndex = uint64(0)
+		}
+
+		var isValidatorExited bool
+		var isValidatorWithdrawn bool
+		index, found := st.ValidatorIndexByPubkey(bytesutil.ToBytes48(pd.PublicKey))
+		if found {
+			val, err := st.ValidatorAtIndexReadOnly(index)
+			if err != nil {
+				return nil, errors.Wrap(err, "could not get validator")
+			}
+			isValidatorExited = val.ExitEpoch() < params.BeaconConfig().FarFutureEpoch
+			isValidatorWithdrawn = val.WithdrawableEpoch() < currentEpoch+1
+		}
+
+		if !isValidatorWithdrawn && !isValidatorExited {
+			isChurnLimitReached := primitives.Gwei(processedAmount+pd.Amount) > availableForProcessing
+			// If churn limit is reached, move to the next epoch.
+			if isChurnLimitReached {
+				currentEpoch += 1
+
+				activeBalance += pendingBalanceToBeActivated[currentEpoch]
+				finalizedSlot += params.BeaconConfig().SlotsPerEpoch
+
+				depBalToConsume = availableForProcessing - primitives.Gwei(processedAmount)
+				// deposit_balance_to_consume only matters when the churn limit is reached.
+				availableForProcessing = depBalToConsume + helpers.ActivationBalanceChurnLimit(activeBalance)
```

Comment:
Future activations can be calculated quite precisely, but as you mentioned, future exits cannot be calculated exactly.

As `pendingBalanceToBeActivated` seems non-intuitive(this is the point I worried about), it would be better to remove this logic and fix `activeBalance`. Thank you!